### ISSUE TITLE
Not usable in SteamPlay/Proton/Wine

### ISSUE THREAD

**Issue by Preisschild** [IDI-Systems/acre2#653]
**Arma 3 Version:** `1.90.145471`
**CBA Version:** `3.10.1`
**ACRE2 Version:** `3.12.6`

**Mods:**
```
- CBA_A3
- acre2
- ACE3 (same without ACE)
```

**Description:**
- Using ACRE2 with Steam's Proton doesn't work correctly.
- When I come to the map-screen of a mission, I get like 0.1 FPS. This continues when I press continue and be in the mission. 

**Steps to reproduce:**
- Use ACRE with Steam' Proton

**Where did the issue occur?**
- Everywhere

**Placed Modules:**
- None

**RPT log file:**
- [gist](https://gist.github.com/Preisschild/bfb5930eeacf2b8d95253db6dbc08a58)

**ACRE log**
```
2019-03-26 22:13:05,600-{INFO }- ACRE Loaded
2019-03-26 22:15:09,446-{INFO }- Generating PBO List
2019-03-26 22:15:09,446-{ERROR}- Error opening object for pbo search
2019-03-26 22:15:09,450-{INFO }- Attempting to load map from PBO...
2019-03-26 22:15:09,450-{ERROR}- WRP unable to find wrp file: \A3\map_Altis\Altis.wrp
2019-03-26 22:15:09,450-{INFO }- ERROR Map Loading, Error Code: -2
2019-03-27 10:28:41,956-{INFO }- ACRE Loaded
2019-03-27 10:29:38,955-{INFO }- Generating PBO List
2019-03-27 10:29:38,955-{ERROR}- Error opening object for pbo search
2019-03-27 10:29:38,958-{INFO }- Attempting to load map from PBO...
2019-03-27 10:29:38,958-{ERROR}- WRP unable to find wrp file: \A3\map_Altis\Altis.wrp
2019-03-27 10:29:38,958-{INFO }- ERROR Map Loading, Error Code: -2
2019-03-27 10:42:23,844-{INFO }- ACRE Loaded
2019-03-27 22:26:49,274-{INFO }- ACRE Loaded
2019-03-27 22:27:45,698-{INFO }- Generating PBO List
2019-03-27 22:27:45,698-{ERROR}- Error opening object for pbo search
2019-03-27 22:27:45,699-{INFO }- Attempting to load map from PBO...
2019-03-27 22:27:45,699-{ERROR}- WRP unable to find wrp file: \A3\map_Altis\Altis.wrp
2019-03-27 22:27:45,700-{INFO }- ERROR Map Loading, Error Code: -2
```

Screenshot of Pop-Up that pops up after starting a mission: https://i.imgur.com/fJCbiAL.jpg

**Comment by jonpas** [comment]
As expected, issue is with PBO opening, due to Windows-specific implementations.

I will try to tackle this eventually, but no ETA.

**Comment by Preisschild** [comment]
Maybe this could get fixed at wine-level. Is there a specific windows dll that handles that?

**Comment by TheMagnetar** [comment]
Not only on a PBO level but also on  shared queues

**Comment by TheMagnetar** [comment]
If this wants to be solved, a valid solution for concurrent queues has to be found.

**Comment by jonpas** [comment]
@Preisschild are you on our Slack? If not, can you join, would be much easier to discuss. :)

**Comment by TheMagnetar** [comment]
It's how the TS plugin is implemented and that is msc c++. Unless that part is changed, nothing can be done

**Comment by jonpas** [comment]
@TheMagnetar you can just run TeamSpeak inside Proton too, as long as required libs are there. But I heard connection to outside works just as well, if we'd compile TS plugin for Linux.

**Comment by TheMagnetar** [comment]
That's the point @jonpas. No concurrent_queue

**Comment by NouberNou** [comment]
Ultimately this should be handled by a Linux specific set of library calls. The stuff being used to get the open PBO file handles is actually not even documented well in the Windows API and is considered "internal" to MS.

**Comment by TacticalTux** [comment]
I am attempting to get ACRE to work as well. I am getting a consistent error of
C0000005 ACCESS_VIOLATION that causes crashing whenever I attempt to load in with ACRE, shortly after "Receiving Data." My acre_dll.log file reads only "ACRE Loaded" and "Generating PBO list", so it does not give me much useful information at the moment.

I am using proton-tkg to use an updated version of Wine in Steam that allows for better DLL compatibility, specifically Proton-tkg 4.12.1.r15.

**Comment by mj-sakellaropoulos** [comment]
I have this mostly working on Proton 5.0 and Ubuntu 19.10.
I installed the TeamSpeak client into the same Proton/wine prefix as Arma3, plugin was automatically copied.

Tested on multiplayer: Was able to Tx/Rx in local chat, able to Tx on radio but could not Rx on radio.
I am getting the exact same .wrp/PBO error.
acre_dll.log: [https://gist.github.com/mj-sakellaropoulos/d553052fa1feb91f458302c849475d1b](https://gist.github.com/mj-sakellaropoulos/d553052fa1feb91f458302c849475d1b)

**Comment by jonpas** [comment]
Nice! Yes, map loading is definitely the blocker here. That's using Windows API, so if someone comes up with a good Proton/Wine solution, we can take a look at integrating it of course.

**Comment by mj-sakellaropoulos** [comment]
Cool. I am looking into the function, If I understand correctly, ```generate_pbo_list``` is trying to get a list of all open file (PBO) handles for the Arma3 process and duplicating those handles to use them? I'm going to try running this with more verbose wine logs and see what I get. From what I can tell the undocumented [SYSTEM_HANDLE_INFORMATION](https://www.geoffchappell.com/studies/windows/km/ntoskrnl/api/ex/sysinfo/handle.htm) is actually implemented in wine.

Update: I dont see anything relevant in the proton logs. And [here](https://github.com/wine-mirror/wine/blob/59987bc9ecdd0dbafd768a95c21a14884bc77c07/include/winternl.h#L1488) is the definition in wine and [here](https://github.com/wine-mirror/wine/blob/59987bc9ecdd0dbafd768a95c21a14884bc77c07/dlls/ntdll/nt.c#L2876) is the implementation.

**Comment by jonpas** [comment]
@NouberNou can maybe get more detailed information, but that looks correct.

**Comment by NouberNou** [comment]
Basically yea, we are using some fairly undocumented Windows API functions to pull in open file handles to find PBOs. That seems to be the part that has always had issues in Wine.

**Comment by mj-sakellaropoulos** [comment]
I narrowed down the issue to the [use](https://github.com/IDI-Systems/acre2/blob/8eb6bad449ba980e1d0b124de150542f5e895c27/extensions/src/ACRE2Arma/common/pbo/search.cpp#L198) of SystemHandleInformationEx instead of SystemHandleInformation. 

I swapped them, now it page faults in the for loop after a couple hundred iterations. I've tried debugging it but winedbg crashes along with arma so I've not gotten very far. 

Here is my working version of search.cpp, it crashes at Line 320.
https://gist.github.com/mj-sakellaropoulos/61f9eb8d8582934552b6b3b974045fd1#file-search-cpp
Here is the log output:
https://gist.github.com/mj-sakellaropoulos/61f9eb8d8582934552b6b3b974045fd1#file-acre_output-txt

Update: Tested my modified acre_x64.dll on windows, it pulls the handles fine and finishes finding all the pbo's. Looks like i'm back to square one with a wine specific problem...

**Comment by jacobeva** [comment]
Any updates on this? Would really like to use ACRE!

**Comment by mj-sakellaropoulos** [comment]
Not really since I havent found a good way to debug this without eventually having to compile a custom version of proton.
There are also other issues in arma for linux, like no thermals and severe stutter, audio spikes...

Basically theres two options:
1. Fix proton to emulate the syscalls properly
2. Somehow bypass pbo unlocking method in acre to read pbo files directly without dealing with syscalls.

**Comment by ninelore** [comment]
@mj-sakellaropoulos the other issues u mentioned are all fixable using winetricks/protontricks
`protontricks 107410 d3dcompiler_43 d3dx10_43 d3dx11_43 xact_x64`

I need to try something. I'll report back

**Comment by dedmen** [comment]
So if you had a way to have the filepaths to all modfolders that Arma is loading, would that fix the problem? ~~Cuz I'm probably adding a command that does that.~~ (Nope, data safety crap)
I heard about "file locking issues" but I cannot ssee anything about that in here?

**Comment by mj-sakellaropoulos** [comment]
> I heard about "file locking issues" but I cannot ssee anything about that in here?

Search the slack for keywords such as "wrp", "pbo", "wine" and "proton".

TLDR: (afaik) Arma locks all the mod PBOs, so external ACRE dll cannot read the map files to get at the terrain data.

**Comment by dedmen** [comment]
> Search the slack for keywords such as "wrp", "pbo", "wine" and "proton".

Well slack people told me to check the github ticket. And due to slack message limit there are only 4 messages and none say anything about being unable to read.

> Arma locks all the mod PBOs, so external ACRE dll cannot read the map files to get at the terrain data.

Arma should lock them yes, but only to prevent write/moving, that shouldn't be blocking reads.

**Comment by ninelore** [comment]
Does ACRE really need `SystemHandleInformationEx` or is `SystemHandleInformation` maybe enough? Because only the latter one is inplemented in wine and should work with some Adjustments.

**Comment by mj-sakellaropoulos** [comment]
> Does ACRE really need `SystemHandleInformationEx`

I dont think so, see my comment above for details. I replaced it in the code and it ran fine on windows but crashed on wine outright.

Someone with more cpp experience could probably pull it off.

**Comment by ArtyomFR** [comment]
You can bypass WINE and play all your Windows game.
WINE can be really tricky and not really reliable on the long therme.
If you really want to play windows games on linux, the most efficient way is to create a KVM virtual machine with full virtio devices and attach your GPU with VGA passthrough. Although this is quiet "easy" on desktop, it is far more difficult even impossible on laptop.
If you managed to succeed, you get a full Windows where you can do whatever you want.
I managed to play Metro Exodus with this configuration without noticing any performance issue between the baremetal Windows and the virtual one. (in fact, this is the same Windows, I just load the barmetal one as a virtual machine on the linux host).

**Comment by jonpas** [comment]
@ArtyomFR that is off-topic.

**Comment by DrymarchonShaun** [comment]
I'm having issues using teamspeak from my arma 3 proton bottle; i'm getting `Failed to connect to server: trying to connect to a server with an unsupported ip protocol` when i try to join a server, but only from the proton version of teamspeak. you guys seem to have gotten it running, do any of you know how to get it to work?

**Comment by ninelore** [comment]
> I'm having issues using teamspeak from my arma 3 proton bottle; i'm getting `Failed to connect to server: trying to connect to a server with an unsupported ip protocol` when i try to join a server, but only from the proton version of teamspeak. you guys seem to have gotten it running, do any of you know how to get it to work?

I suggest u to join the ArmaOnUnix Discord https://discord.gg/p28Ra36

**Comment by DrymarchonShaun** [comment]
> > I'm having issues using teamspeak from my arma 3 proton bottle; i'm getting `Failed to connect to server: trying to connect to a server with an unsupported ip protocol` when i try to join a server, but only from the proton version of teamspeak. you guys seem to have gotten it running, do any of you know how to get it to work?
> 
> I suggest u to join the ArmaOnUnix Discord https://discord.gg/p28Ra36

joined, thanks. no idea that existed.

**Comment by ninelore** [comment]
~~someone on the ArmaOnUnix Discord just claimed to have it working fine (on Proton 5.21-GE-1). Can anyone reproduce as i dont have access to my main PC atm?~~

see next comment

**Comment by ninelore** [comment]
I just solved the Problem!

SystemHandleInformationEx got implemented into wine staging patchset     
Proton ships without staging patches

--> use a custom Proton!     
My recommendation: [proton-tkg-git](https://github.com/Frogging-Family/wine-tkg-git)

**Comment by ninelore** [comment]
## [My Guide is up](https://armaworld.de/forum/thread/4992-ger-eng-arma-on-linux-ninelore-s-kollektiver-guide-inkl-acre/?postID=39909#post39909)
(The ACRE part will be added shortly)

**Comment by ninelore** [comment]
ACRE Part in the Guide is out
It is including a Teamspeak Bash Script and a prebuilt patched proton

I consider this Issue solved ;)

**Comment by jonpas** [comment]
@ninelore Awesome work as I already told you on Slack, as well as to anyone else involved with the relevant projects.

If you would like to make your guide more accessible, we would not mind if you throw a link onto our documentation (noting it is 3rd party as we cannot officially support this at this time). ;)
